ampd/src/grpc/event_filters.rs

```diff
@@ -0,0 +1,380 @@
+use std::str::FromStr;
+
+use axelar_wasm_std::nonempty;
+use cosmrs::AccountId;
+use error_stack::{report, Report, Result};
+use report::ResultCompatExt;
+use thiserror::Error;
+use tonic::Status;
+
+use crate::types::TMAddress;
+use crate::{event_sub, PREFIX};
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("empty filter")]
+    EmptyFilter,
+    #[error("invalid contract address in filter")]
+    InvalidContractAddress(String),
+}
+
+impl From<&Error> for Status {
+    fn from(error: &Error) -> Self {
+        match error {
+            Error::EmptyFilter => Status::invalid_argument("empty filter provided"),
+            Error::InvalidContractAddress(contract) => Status::invalid_argument(format!(
+                "invalid contract address {} provided in filters",
+                contract
+            )),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct EventFilter {
+    event_type: Option<nonempty::String>,
+    contract: Option<TMAddress>,
+}
+
+impl TryFrom<ampd_proto::EventFilter> for EventFilter {
+    type Error = Report<Error>;
+
+    fn try_from(event_filter: ampd_proto::EventFilter) -> Result<Self, Error> {
+        let event_type = event_filter.r#type.try_into().ok();
+        let contract = if event_filter.contract.is_empty() {
+            None
+        } else {
+            let contract = AccountId::from_str(&event_filter.contract)
+                .change_context(Error::InvalidContractAddress(event_filter.contract.clone()))
+                .and_then(|contract| {
+                    if contract.prefix() != PREFIX {
+                        return Err(report!(Error::InvalidContractAddress(
+                            event_filter.contract
+                        )));
+                    }
+
+                    Ok(contract)
+                })
+                .map(Into::into)?;
+
+            Some(contract)
+        };
+
+        if event_type.is_none() && contract.is_none() {
+            return Err(report!(Error::EmptyFilter));
+        }
+
+        Ok(EventFilter {
+            event_type,
+            contract,
+        })
+    }
+}
+
+impl EventFilter {
+    pub fn filter(&self, event_type: &str, contract: &Option<TMAddress>) -> bool {
+        if self
+            .event_type
+            .as_ref()
```

Comment:
why is as_ref necessary?

Reply:
becuz `is_some_and` consumes `self` for some reason.

Reply:
ok that's dumb